src/backend/dx12/src/command.rs

```diff
@@ -24,29 +24,89 @@ fn get_rect(rect: &target::Rect) -> winapi::D3D12_RECT {
 pub struct RenderPassCache {
     render_pass: n::RenderPass,
     frame_buffer: n::FrameBuffer,
-    next_subpass: usize,
     render_area: winapi::D3D12_RECT,
     clear_values: Vec<ClearValue>,
 }
 
 #[derive(Clone)]
 pub struct CommandBuffer {
-    pub(crate) raw: ComPtr<winapi::ID3D12GraphicsCommandList>,
-    pub(crate) allocator: ComPtr<winapi::ID3D12CommandAllocator>,
+    raw: ComPtr<winapi::ID3D12GraphicsCommandList>,
+    allocator: ComPtr<winapi::ID3D12CommandAllocator>,
 
     // Cache renderpasses for graphics operations
-    pub(crate) pass_cache: Option<RenderPassCache>,
+    pass_cache: Option<RenderPassCache>,
+    next_subpass: usize,
 }
 
 unsafe impl Send for CommandBuffer { }
 
 impl CommandBuffer {
+    pub(crate) fn new(
+        raw: ComPtr<winapi::ID3D12GraphicsCommandList>,
+        allocator: ComPtr<winapi::ID3D12CommandAllocator>,
+    ) -> Self {
+        CommandBuffer {
+            raw,
+            allocator,
+            pass_cache: None,
+            next_subpass: !0,
+        }
+    }
+
+    pub(crate) unsafe fn as_raw_list(&self) -> *mut winapi::ID3D12CommandList {
+        self.raw.as_mut() as *mut _ as *mut _
+    }
+
+    fn insert_subpass_barriers(&self, subpass: SubpassRef) {
+        let state = self.pass_cache.as_ref().unwrap();
+
+        let mut transition_barriers = Vec::new();
+        for dependency in &state.render_pass.dependencies {
+            if dependency.passes.end != subpass {
+                continue
+            }
+            for (i, att) in state.render_pass.attachments.iter().enumerate() {
+                let state_src = conv::map_image_resource_state(dependency.accesses.start, att.layouts.start);
+                let state_dst = conv::map_image_resource_state(dependency.accesses.end, att.layouts.end);
+
+                if state_src == state_dst {
+                    continue;
+                }
+
+                let resource = match state.frame_buffer.color.get(i) {
+                    Some(view) => view.resource,
+                    None => state.frame_buffer.depth_stencil[0].resource,
```

Comment:
We can't have multiple ds views in a framebuffer?

Reply:
we should have, this will be TODO, esp considering that even a single depth buffer is not supported by our rough backends yet